cardano-db/src/Cardano/Db/Insert.hs

```diff
@@ -142,21 +172,70 @@ data DbInsertException
 
 instance Exception DbInsertException
 
--- | Insert a record (with a Unique constraint), and return 'Right key' if the
--- record is inserted and 'Left key' if the record already exists in the DB.
-insertByReturnKey
+-- Insert without checking uniqueness constraints. This should be safe for most tables
+-- even tables with uniqueness constraints, especially block, tx and many others, where
+-- uniqueness is enforced by the ledger.
+insertUnchecked
     :: ( AtLeastOneUniqueKey record
        , MonadIO m
        , MonadBaseControl IO m
        , PersistEntityBackend record ~ SqlBackend
        )
     => String -> record -> ReaderT SqlBackend m (Key record)
-insertByReturnKey vtype value = do
-    res <- getByValue value
-    case res of
-      Nothing -> handle exceptHandler $ insert value
-      Just r -> pure $ entityKey r
+insertUnchecked vtype =
+    handle exceptHandler . insert
   where
     exceptHandler :: MonadIO m => SqlError -> ReaderT SqlBackend m a
     exceptHandler e =
       liftIO $ throwIO (DbInsertException vtype e)
+
+-- Insert, getting PostgreSQL to check the uniqueness constaint, and if it is violated, rewrite
+-- the first field with the same value to force PostgresSQL to return the row identifier.
+insertCheckUnique
+    :: forall m record.
+        ( MonadBaseControl IO m
+        , MonadIO m
+        , PersistEntity record
+        , PersistRecordBackend record SqlBackend
+        )
+    => String -> record -> ReaderT SqlBackend m (Key record)
+insertCheckUnique vtype record = do
+    res <- handle exceptHandler $ rawSql query values
+    case res of
+      [ident] -> pure ident
+      _other -> error $ mconcat [ "insertByReturnKey2: Inserting ", vtype, " failed with ", show res ]
+  where
+    query :: Text
+    query =
+      Text.concat
+        [ "INSERT INTO "
+        , unDBName (entityDB . entityDef $ Just record)
+        , " (", Util.commaSeparated fieldNames
+        , ") VALUES (", Util.commaSeparated placeholders
+        , ") ON CONFLICT ON CONSTRAINT "
+        , unDBName (uniqueDBName . head . entityUniques . entityDef $ Just record)
+        , " DO UPDATE SET ", head fieldNames, " = ", head placeholders
+        , " RETURNING id ;"
+        ]
+
+    values :: [PersistValue]
+    values = pvalues ++ [head pvalues]
+
+    pvalues :: [PersistValue]
+    pvalues = map toPersistValue (toPersistFields record)
+
+    fieldNames, placeholders :: [Text]
+    (fieldNames, placeholders) =
+      unzip (Util.mkInsertPlaceholders (entityDef (Proxy @record)) escape)
+
+    exceptHandler :: SqlError -> ReaderT SqlBackend m a
+    exceptHandler e =
+      liftIO $ throwIO (DbInsertException vtype e)
+
+escape :: DBName -> Text
```

Comment:
A pity this is not exported

Reply:
Yes, and it seems it is changing between `persistent-2.11.0.1` and the next release.